### PR TITLE
Remove usage of OkHttp

### PR DESCRIPTION
**Pull request by Yoshani**

### CODE REVIEW

File: src/main/java/io/scim2/swagger/client/ScimApiClient.java [hunk 1]
```diff
@@ -105,17 +137,22 @@ public class ScimApiClient {
     private boolean verifyingSsl;
     private KeyManager[] keyManagers;
 
-    private OkHttpClient httpClient;
+    private HttpClient httpClient;
+    private final HttpClientBuilder httpClientBuilder;
+    private RequestConfig requestConfig;
```
Comment (bhagyasakalanka) [inline src/main/java/io/scim2/swagger/client/ScimApiClient.java:142]: Is there any special reason to keep `httpClientBuilder` and `requestConfig` as class level attributes? Feels like, we should define these inside the method level when initiating the `httpClient`

File: src/main/java/io/scim2/swagger/client/ScimApiClient.java [hunk 2]
```diff
@@ -758,23 +902,25 @@ public String escapeString(String str) {
...
+     * @param response   HTTP response
      * @param returnType The type of the Java object
      * @return The deserialized Java object
      * @throws ScimApiException If fail to deserialize response body, i.e. cannot read response body
-     *   or the Content-Type of the response is not supported.
+     *                          or the Content-Type of the response is not supported.
      */
     @SuppressWarnings("unchecked")
-    public <T> T deserialize(Response response, Type returnType) throws ScimApiException {
+    public <T> T deserialize(HttpResponse response, Type returnType) throws ScimApiException {
```
Comment (bhagyasakalanka) [inline src/main/java/io/scim2/swagger/client/ScimApiClient.java:913]: Changing the signature of the public method is not a good idea. Hope we are going to release a major version of scim?

File: src/main/java/io/scim2/swagger/client/ScimApiClient.java [hunk 3]
```diff
@@ -973,143 +1125,165 @@ public <T> T handleResponse(Response response, Type returnType) throws ScimApiEx
...
+            entityRequest.setEntity(reqBody);
         }
-        Request request = reqBuilder.method(method, reqBody).build();
         return request;
     }
 
+    private HttpUriRequest createHttpRequest(String method, String url) {
+
+        switch (method) {
+            case "GET":
```
Comment (bhagyasakalanka) [inline src/main/java/io/scim2/swagger/client/ScimApiClient.java:1214]: We could define a enum for this? Or constants. Also check if we have any other library which provide these methods as constants?

File: src/main/java/io/scim2/swagger/client/ScimApiClient.java (hunk 2 above)
Comment (Yoshani) [reply src/main/java/io/scim2/swagger/client/ScimApiClient.java:913]: Yes a major version of the scim2 client will be released.

File: src/main/java/io/scim2/swagger/client/ScimApiClient.java (hunk 1 above)
Comment (Yoshani) [reply src/main/java/io/scim2/swagger/client/ScimApiClient.java:142]: These were introduced as class level attributes to support setting and getting some configurations which cannot be set or retrieved directly from the `httpClient` object in the Apache library.
`requestConfig` - used in `getConnectTimeout` method to get the timeout value being set in `setConnectTimeout` method.
`httpClientBuilder` - used for setting configs like setting timeout, setting SSLHostnameVerifier. HttpClient is built after setting these. 

Class level variables are required to support these settings being set in different methods.

File: src/main/java/io/scim2/swagger/client/ScimApiClient.java [hunk 4]
```diff
@@ -973,143 +1127,169 @@ public <T> T handleResponse(Response response, Type returnType) throws ScimApiEx
...
-                reqBody = null;
+        HttpEntity reqBody;
+        if (request instanceof HttpEntityEnclosingRequestBase) {
+            HttpEntityEnclosingRequestBase entityRequest = (HttpEntityEnclosingRequestBase) request;
+            if (URL_ENCODED_CONTENT_TYPE.equals(contentType)) {
+                reqBody = buildRequestBodyFormEncoding(formParams);
+            } else if (MULTIPART_FORM_DATA_CONTENT_TYPE.equals(contentType)) {
+                reqBody = buildRequestBodyMultipart(formParams);
+            } else if (body == null) {
+                // use an empty request body (for POST, PUT, and PATCH)
```
Comment (bhagyasakalanka) [inline src/main/java/io/scim2/swagger/client/ScimApiClient.java:1202]: Format the comments. Start with a capital letter and end with a full stop.

File: pom.xml [hunk 5]
```diff
@@ -139,7 +149,12 @@
                             org.wso2.scim2.*;version=${project.version},
                             io.scim2.swagger.client.*;version=${project.version},
                             com.squareup.okio.*,
-                            com.squareup.okhttp.*,
+                            org.apache.http.auth.*; version="${httpclient.orbit.version}",
+                            org.apache.http.client.*; version="${httpclient.orbit.version}",
+                            org.apache.http.conn.*; version="${httpclient.orbit.version}",
+                            org.apache.http.impl.client; version="${httpclient.orbit.version}",
+                            org.apache.http.*;version="${httpcore.orbit.version}",
```
Comment (chamathns) [inline pom.xml:156]: check if we can remove the parent package import

File: src/main/java/io/scim2/swagger/client/ScimApiClient.java [hunk 6]
```diff
@@ -536,39 +610,98 @@ public boolean isDebugging() {
...
+                            (request, context) -> logger.debug(getRequestInfo(request));
+                    HttpResponseInterceptor responseInterceptor =
+                            (response, context) -> logger.debug(getResponseInfo(response));
+
+                    httpClient = HttpClients.custom()
+                            .addInterceptorFirst(requestInterceptor)
+                            .addInterceptorFirst(responseInterceptor)
+                            .build();
+                } else {
+                    logger.info("Disabling debug mode");
```
Comment (chamathns) [inline src/main/java/io/scim2/swagger/client/ScimApiClient.java:630]: do we need this log?

File: src/main/java/io/scim2/swagger/client/ScimApiClient.java [hunk 7]
```diff
@@ -536,39 +610,98 @@ public boolean isDebugging() {
...
+    private String getRequestInfo(HttpRequest request) throws IOException {
+
+        String message = "\nRequest - "
+                + request.getRequestLine().getMethod() + " "
+                + request.getRequestLine().getUri()
+                + "\nHeaders: ["
+                + Arrays.stream(request.getAllHeaders())
+                .map(header -> header.getName() + ": " + header.getValue())
+                .collect(Collectors.joining(", "))
+                + "]";
```
Comment (chamathns) [inline src/main/java/io/scim2/swagger/client/ScimApiClient.java:651]: let's use a string builder

Comment (chamathns) [inline src/main/java/io/scim2/swagger/client/ScimApiClient.java:651]: we can extract this out to a separate method. to improve readability

private String getHeadersInfo(HttpRequest request) {
    return Arrays.stream(request.getAllHeaders())
            .map(header -> header.getName() + ": " + header.getValue())
            .collect(Collectors.joining(", "));
}

File: src/main/java/io/scim2/swagger/client/ScimApiClient.java [hunk 8]
```diff
@@ -536,39 +610,98 @@ public boolean isDebugging() {
...
+
+        if (request instanceof HttpEntityEnclosingRequest) {
+            HttpEntity entity = ((HttpEntityEnclosingRequest) request).getEntity();
+            ByteArrayOutputStream bs = new ByteArrayOutputStream();
+            entity.writeTo(bs);
+            message += "\nPayload:\n" + bs;
+        }
```
Comment (chamathns) [inline src/main/java/io/scim2/swagger/client/ScimApiClient.java:658]: we can extract this to a private method as well.

File: src/main/java/io/scim2/swagger/client/ScimApiClient.java [hunk 9]
```diff
@@ -536,39 +610,98 @@ public boolean isDebugging() {
...
+    private String getResponseInfo(HttpResponse response) throws IOException {
+
+        String message = "\nResponse - "
+                + response.getStatusLine().getStatusCode() + " "
+                + response.getStatusLine().getReasonPhrase()
+                + "\nHeaders: ["
+                + Arrays.stream(response.getAllHeaders())
+                .map(header -> header.getName() + ": " + header.getValue())
+                .collect(Collectors.joining(", "))
+                + "]";
```
Comment (chamathns) [inline src/main/java/io/scim2/swagger/client/ScimApiClient.java:672]: refer above comments for the getRequestInfo method

File: src/main/java/io/scim2/swagger/client/ScimApiClient.java [hunk 10]
```diff
@@ -1256,4 +1459,18 @@ private KeyStore newEmptyKeyStore(char[] password) throws GeneralSecurityExcepti
...
+        Header[] headers = response.getAllHeaders();
+        Map<String, List<String>> extractedHeaders = new HashMap<>();
+
+        for (Header header : headers) {
+            String name = header.getName();
+            String value = header.getValue();
+            extractedHeaders.computeIfAbsent(name, k -> new ArrayList<>()).add(value);
+        }
+
+        return extractedHeaders;
```
Comment (chamathns) [inline src/main/java/io/scim2/swagger/client/ScimApiClient.java:1474]: can use streams to improve readability :)

```suggestion
        return Arrays.stream(response.getAllHeaders())
                 .collect(Collectors.groupingBy(
                     Header::getName, 
                     Collectors.mapping(Header::getValue, Collectors.toList())
                 ));
```

File: pom.xml (hunk 5 above)
Comment (Yoshani) [reply pom.xml:156]: httpcore files are included in the package `org.apache.http`. So this exports only the related content in httpcore.

**Review by chamathns — DISMISSED** [review]

**Review by chamathns — APPROVED** [review]

**Pull request merged**